libraries/libstratosphere/source/settings/impl/settings_key_value_store.cpp

```diff
@@ -0,0 +1,1629 @@
+                    system_data_id = ncm::SystemDataId::PlatformConfigIcosaMariko;
+                    break;
+                case SplHardwareType_Copper:
+                    system_data_id = ncm::SystemDataId::PlatformConfigCopper;
+                    break;
+                case SplHardwareType_Hoag:
+                    system_data_id = ncm::SystemDataId::PlatformConfigHoag;
+                    break;
+                case SplHardwareType_Calcio:
+                    system_data_id = ncm::SystemDataId::PlatformConfigCalcio;
+                    break;
+                case SplHardwareType_Aula:
+                    system_data_id = ncm::SystemDataId::PlatformConfigAula;
+                    break;
+                AMS_UNREACHABLE_DEFAULT_CASE();
+            }
+
+            /* Get the platform configuration system data. */
+            SystemData *system_data = nullptr;
+            R_TRY(GetSystemData<SystemDataTag::PfCfg>(std::addressof(system_data), system_data_id));
+            AMS_ASSERT(system_data != nullptr);
+
+            /* Read the data. */
+            return ReadAllBytes<SystemData>(system_data, out_count, out_buffer, out_buffer_size);
+        }
+
+        Result ReadSystemSaveData(u64 *out_count, char * const out_buffer, size_t out_buffer_size) {
+            /* Check preconditions. */
+            AMS_ASSERT(out_count != nullptr);
+            AMS_ASSERT(out_buffer);
+        
+            /* Get the system save data. */
+            SystemSaveData *system_save_data = nullptr;
+            if (const auto result = GetSystemSaveData(std::addressof(system_save_data), false); R_FAILED(result)) {
+                *out_count = 0;
+                return ResultSuccess();
+            }
+            AMS_ASSERT(system_data != nullptr);
+
+            /* Read the data. */
+            return ReadAllBytes<SystemSaveData>(system_save_data, out_count, out_buffer, out_buffer_size);
+        }
+
+        Result SaveKeyValueStoreMap(const Map &map) {
+            /* Get the system save data. */
+            SystemSaveData *system_save_data = nullptr;
+            R_TRY(GetSystemSaveData(std::addressof(system_save_data), false));
+            AMS_ASSERT(system_save_data != nullptr);
+
+            /* Save the current values of the key value store map. */
+            return SaveKeyValueStoreMapCurrent<SystemSaveData>(*system_save_data, map);
+        }
+
+        template<typename T, typename F>
+        Result SaveKeyValueStoreMap(T &data, const Map &map, F test) {
+            /* Check preconditions. */
+            AMS_ASSERT(test != nullptr);
+
+            /* Create the save data if necessary. */
+            R_TRY_CATCH(data.Create(512_KB)) {
+                R_CATCH(fs::ResultPathAlreadyExists) { /* ... */ }
+            } R_END_TRY_CATCH;
+
+            {
+                /* Open the save data for writing. */
+                R_TRY(data.OpenToWrite());
+                ON_SCOPE_EXIT {
```

Comment:
add NOTE comment that Nintendo only does this if SetFileSize succeeds, or change to match what Nintendo does.